### ISSUE TITLE
[circle+] Formula to calculate gradient

### ISSUE THREAD

**Issue by jyoungyun** [Samsung/ONE#11248]
## Purpose

This issue is where we record what formula we used when calculating the gradient of backward propagation. Based on the formula here, we will implement a gradient, and for details of the implementation, please refer to the source code.

**Comment by jyoungyun** [comment]
# FullyConnected

## formula to calculate gradient
Input : $X$
Weights $W$
Output : $Z$
Bias : $B$

### FullyConnected Forward
##### Formula
$Z = XW + B$

##### Fullyconnected Formula
fullyconnected's `weights` is already transposed. So $W$ here is transposed of the above $W$.
$Z = fc(X, W)$

### FullyConnected Backward
#### Input
##### Formula
$\frac{\partial L}{\partial X}$ = $\frac{\partial L}{\partial Z} W^T $

##### Fullyconnected Formula
$\frac{\partial L}{\partial X}$ =  fc($\frac{\partial L}{\partial Z}, W^T) $

#### Weights
##### Formula
$\frac{\partial L}{\partial W} = (\frac{\partial L}{\partial Z})^T X$

##### Fullyconnected Formula
$\frac{\partial L}{\partial W} = fc((\frac{\partial L}{\partial Z})^T, X^T)$

#### Bias
##### Formula
$$
\begin{flalign}
\frac{\partial L}{\partial B} = 
\begin{bmatrix}
    \frac{\partial L}{\partial z_{11}} + \frac{\partial L}{\partial z_{21}} \\
    \frac{\partial L}{\partial z_{12}} + \frac{\partial L}{\partial z_{22}} \\
\end{bmatrix} 
\end{flalign}
$$

_This code is copied from #11204._

**Comment by jyoungyun** [comment]
## MSE (Mean Square Error)

### Forward

### $\text{Error}(y, y_{t}) = \cfrac{1}{N} \sum_{\text{i=1 to N}} (y_{i} - y_{it})^2$

### Backward

### $\cfrac{\partial E}{\partial y_{i}} = \cfrac{\partial}{\partial y_{i}} \cfrac{1}{N} (y_{i} - y_{it})^2 = \cfrac{\partial}{\partial y_{i}} \cfrac{1}{N} (y_{i}^{2} - 2y_{i} y_{it} + y_{it}^2) = \cfrac{1}{N} (2y_{i} - 2y_{it}) = \cfrac{1}{N} (-2(y_{it} - y_{i})) = -\cfrac{2}{N}(y_{it}-y_{i})$

### $\cfrac{\partial E}{\partial y_{i}} = -\cfrac{2}{N} (y_{it} - y_{i})$

**Comment by zetwhite** [comment]
## MaxPool

**work in progress** 

* Input = $x_1, x_2, x_3 ... x_n$ 

### Forward

$$
\begin{equation}
     MaxPool(x_1, \cdots, x_n) = max(x_1, \cdots, x_n) =
         \begin{cases}
             x_1 & \text{if } x_1 > x_2, x_1 > x_3, \cdots x_1 > x_n\\
             x_2 & \text{if } x_2 > x_1, x_2 > x_3, \cdots x_2 > x_n\\ 
             \vdots & \vdots \quad \vdots \qquad \vdots \qquad \vdots \qquad \vdots \quad\vdots\\
             x_n & \text{if } x_n > x_1, x_n > x_3, \cdots x_n > x_{n - 1}
         \end{cases} 
\end{equation}
$$


### Backward

$$
\begin{equation}     
\frac{\partial{MaxPool}}{\partial{x_i}} =
         \begin{cases}
             1 & \text{if } x_i > x_2, x_i > x_3, \cdots x_i > x_n\\
             0 & \text{otherwise},         	 
         \end{cases} 
\end{equation}
$$

### Notes

**Comment by glistening** [comment]
@zetwhite MaxPool Forward formula does not handle the case `=` (e.g. $x_1$ = $x_2$ = ... = $x_n$). Could you leave the reference where you get this formula?

**Comment by jyoungyun** [comment]
## Softmax

### Forward

```math
S(x_{i}) = \frac{\exp{x_{i}}}{ \sum_{k=1} \exp{x_{j}} }
```

### Backward

```math
\frac{\partial S(x_{i})}{\partial x} = \begin{cases}
& S(x_i) \cdot (1 - S(x_i)) \text{  if i = j }  \\
& -S(x_i) \cdot S(x_j) \text{  if i != j }  \end{cases}
```

**Comment by ragmani** [comment]
## Conv2D
### Forward

$$
\begin{align}
Z_{i,j,k}^{(m_i)} &= \sum\limits_{a} \sum\limits_{b} \sum\limits_{c} X_{S_i * i + a - P_t, S_j * j + b - P_l, c}^{(m_i)} * W_{a,b,c,k}  + b_k, \\
& \text{if } 0 <= S_i * i + a - P_t < \text{input height}, 0 <=  S_j * j + b - P_l < \text{input width}
\end{align}
$$

$a$ : weight height
$b$ : weight width
$c$ : input channel
$k$ : output channel
$i$ : output height
$j$ : output width
$m_i$ : $i$ th batch

$S_i$ : Stride height
$S_j$ : Stride width 
$P_t$ : Pad top
$P_l$ : Pad left

### Backward
#### Input
$$
\begin{align}
\frac{\partial L}{\partial X_{S_i * i + a - P_t, S_j * j + b - P_l, c}^{m_i}}&= \sum\limits_{k} \sum\limits_{a} \sum\limits_{b} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}} W_{a,b,c,k} , \\
& \text{if } 0 <= S_i * i + a - P_t < \text{input height}, 0 <=  S_j * j + b - P_l < \text{input width}
\end{align}
$$

#### Weights
$$
\begin{align}
\frac{\partial L}{\partial W_{a,b,c,k}}  &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}} X_{S_i * i + a - P_t, S_j * j + b - P_l, c}^{m_i} , \\
& \text{if } 0 <= S_i * i + a - P_t < \text{input height}, 0 <=  S_j * j + b - P_l < \text{input width}
\end{align}
$$


#### Bias
$$
\begin{align}
\frac{\partial L}{\partial b_k} &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}}  \frac{\partial Z_{i,j,c}^{m_i}}{\partial b_k} \\
    &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}} {\cdot } 1 \\
    &= \sum\limits_{m} \sum\limits_{i} \sum\limits_{j} \frac{\partial L}{\partial Z_{i,j,k}^{m_i}}
\end{align}
$$

**Comment by jyoungyun** [comment]
## Categorical Cross Entropy

### Forward

```math
Error(y, y_t) = - \sum_{i=1}^N y_{it} log(y_i)
```

### Backward

```math
\frac{\partial E}{\partial y_i} = - \frac{y_{it}}{y_i}
```

**Comment by jyoungyun** [comment]
## ReduceMean

input: X of size (C, H, W)
keep_dims: false
output: (C, 1, 1)

### Forward

```math
y[c, 0, 0] = \frac{1}{H*W} * sum(X[c, i, j])
```

### Backward

```math
\frac{\partial L}{\partial X[c, i, j]} =  \frac{1}{H*W} * \frac{\partial L}{\partial Y[c, 0, 0]}
```

**Comment by icodo98** [comment]
## Average Poll

- \( X \) is the input feature map.
- \( Y \) is the output feature map (pooled output).
- \( k * k \) is the size of the pooling window.
- \( s \) is the stride (how much the window moves after each pooling).
- \( i, j \) are the indices of the output map.
- \( m, n \) are the indices within the pooling window.

### Forward

```math
Y(i, j) = \frac{1}{k \times k} \sum_{m=0}^{k-1} \sum_{n=0}^{k-1} X(i \cdot s + m, j \cdot s + n)
```

### Backward

```math
\frac{\partial L}{\partial X(i \cdot s + m, j \cdot s + n)} = \frac{1}{k \times k} \frac{\partial L}{\partial Y(i, j)}
```
